### DISCUSSION TITLE
Cannot upload any images

### DISCUSSION THREAD

**Discussion by SangeloDev** [outline/outline#3307] [Self Hosting] [answered]
EDIT: Answer to the problem can be found [here](https://github.com/outline/outline/discussions/3307#discussioncomment-2483879). Thanks to @almereyda for helping out!

When trying to upload a team image in Details, I always get this error:
<img width="473" alt="image" src="https://user-images.githubusercontent.com/47925793/160909035-19cbca2f-e5cf-46c8-b5cb-5c012690b6fa.png">

This is what my config looks like:
```AWS_ACCESS_KEY_ID=<secret>
AWS_SECRET_ACCESS_KEY=<secret>
MINIO_ACCESS_KEY=<secret>
MINIO_SECRET_KEY=<secret>
AWS_REGION=eu-central-1
AWS_S3_UPLOAD_BUCKET_URL=http://XXX.XX.XXX.XXX:9001/outline
AWS_S3_UPLOAD_BUCKET_NAME=outline
AWS_S3_UPLOAD_MAX_SIZE=26214400
AWS_S3_FORCE_PATH_STYLE=true
AWS_S3_ACL=private
```

And this is the console output I am seeing:
```info: [http]   <-- POST /api/attachments.create
info: [http]   --> POST /api/attachments.create 200 24ms 1005b
info: [processor] backlinks processing attachments.create
info: [processor] debouncer processing attachments.create
info: [processor] imports processing attachments.create
info: [processor] exports processing attachments.create
info: [processor] revisions processing attachments.create
info: [processor] notifications processing attachments.create
info: [processor] slack processing attachments.create
```

I don't know how to continue. Any help would be appreciated.

**Comment by almereyda** [comment, marked as answer]
Could you try opening the network tab of your browser's developer tools to check if you can catch the request that fails? Eventually we learn other error messages there.

**Reply by SangeloDev** [reply]
There's nothing there.
I opened the page, opened the network tab, and clicked reload.
I got a normal request as seen here:
<img width="409" alt="image" src="https://user-images.githubusercontent.com/47925793/161035934-50a5f52d-b180-4291-b1f0-52e880fddd49.png">
I then proceeded to upload the avatar, but nothing came up, only that Outline error.
I also have problems uploading any other file, like images inside of documents, etc.

**Reply by almereyda** [reply]
Could you try setting the variables

https://github.com/outline/outline/blob/fea5f69a38d19244e70cac2e74e260cb84496a31/.env.sample#L38-L40

In the configuration that you pasted some Minio credentials were present, which are not read by Outline, and also an `AWS_SECRET_ACCESS_KEY`. I suspect that when you also set the `AWS_ACCESS_KEY_ID`, the connection would work.

Eventually you could also try increasing the Outline loglevel with setting `DEBUG` to

    DEBUG=cache,presenters,events,emails,mailer,utils,http,server,services

and see if that gives you more details about what errors out. I'm using all the possible settings here, since I do not know right now where S3 related errors would turn up.

**Reply by SangeloDev** [reply]
So I'd have to get an AWS server, or am I not understanding properly?
I've set my MinIO secret and access keys under AWS_* because I didn't understand where to put them elsewhere.

I'll also try with the log levels increased later, because I don't have much time now.

EDIT: There isn't happening much with the DEBUG level increased. Still the same logs.

**Reply by almereyda** [reply]
No, you do not need AWS. You can use Minio. This is also what is present in the official installation instructions.

https://app.getoutline.com/share/770a97da-13e5-401e-9f8a-37949c19f97e/doc/docker-7pfeLP5a8t

This also mentions to consult the sample environment file for all settings. This section is especially for configuring S3 storage:

https://github.com/outline/outline/blob/fea5f69a38d19244e70cac2e74e260cb84496a31/.env.sample#L35-L46

There is a page about how to set up Amazon S3 at https://wiki.generaloutline.com/share/125de1cc-9ff6-424b-8415-0d58c809a40f

As this does not explain how to configure Minio, which is used in the Docker Compose example, the configuration of S3 is left to the community. A few days ago I had left some comments in https://github.com/outline/outline/discussions/3281#discussioncomment-2430516 how we managed to create a working Outline instance with S3.

An example of a working Compose YAML with environment had been posted to https://github.com/outline/outline/discussions/3291#discussioncomment-2460649

**Reply by SangeloDev** [reply]
I think I managed to proceed further, now I get new errors.
Turns out the first problem was that I connected over http, not https.
But now I am getting a ``403 Forbidden`` error when trying to upload:
![Screenshot](https://user-images.githubusercontent.com/47925793/161089354-c0cb2fac-0bf0-4f27-893f-964f0c640fb4.png)

**Reply by almereyda** [reply]
Okay, from here I do not know any further. I thought uploading is handled transparently through the Outline backend, but in testing this with an image in a document, I could replicate that POST request.

In my case it is behind an HTTP/2 proxy, and not an HTTP/3 one. It could be the documented `https-portal`, which I have never touched. Are you using the Minio setup from the Outline distribution, or another one?

You could try to review our Minio settings, which I had posted in https://github.com/outline/outline/discussions/3281#discussioncomment-2430516, and esp. check for the correct application of policy, user and bucket through `mc`.

**Reply by SangeloDev** [reply]
> Are you using the Minio setup from the Outline distribution, or another one?

I am hosting Outline from source, I don't use the docker image for that.

> You could try to review our Minio settings, which I had posted in https://github.com/outline/outline/discussions/3281#discussioncomment-2430516, and esp. check for the correct application of policy, user and bucket through `mc`.

I followed these instructions, doesn't seem to make a difference. The above happens with the policies applied.

**Reply by SangeloDev** [reply]
Actually, I might've found something here:
```
XHRPOSThttps://minio.<domain>/outline
[HTTP/3 403 Forbidden 66ms]

<?xml version="1.0" encoding="UTF-8"?>

<Error><Code>InvalidAccessKeyId</Code><Message>The Access Key Id you provided does not exist in our records.</Message><BucketName>outline</BucketName><Resource>/outline</Resource><RequestId>xxxxx</RequestId><HostId>xxxxx</HostId></Error>
```
Maybe I should try changing the secret keys, but I don't know with what?

**Reply by dveretennikov1** [reply]
You have to use HTTPS with valid ssl-cert in
AWS_S3_UPLOAD_BUCKET_URL

**Reply by almereyda** [reply]
> Maybe I should try changing the secret keys, but I don't know with what?

Please check the [MinIO | MinIO Client Quickstart Guide](https://docs.min.io/docs/minio-client-quickstart-guide.html).

In https://github.com/outline/outline/discussions/3281#discussioncomment-2430516 we have documented the steps to configure Minio with the `mc` (Minio Client) command. Maybe they together with the documentation above can give you an idea how to create dedicated credentials for your bucket?

The individual subcommands are explained more thoroughly in https://docs.min.io/docs/minio-client-complete-guide.html under section **7. Commands**.

**Reply by SangeloDev** [reply]
I searched through all the documentation of MinIO and ended up finding out that ``AWS_ACCESS_KEY_ID=`` should equal to the username of the user you add and set the policy described [here](https://github.com/outline/outline/discussions/3281#discussioncomment-2430516), in my case it's ``AWS_ACCESS_KEY_ID=outline`` and the secret is the password of that user. 
That ended up fixing everything! Now I can finally upload images. Thanks, @almereyda, for helping out! I really appreciate that.

**Reply by almereyda** [reply]
Perfect.
